core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,437 @@
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super T> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+
+        final SurroundingSubscriber<T> surroundingSubscriber = new SurroundingSubscriber<>(
+                head, publisher, tail, subscriber, executor, completionFuture, options);
+        this.surroundingSubscriber = surroundingSubscriber;
+        subscriber.onSubscribe(surroundingSubscriber);
+
+        // To make sure to close the SurroundingSubscriber when this is aborted.
+        if (completionFuture.isCompletedExceptionally()) {
+            completionFuture.exceptionally(cause -> {
+                surroundingSubscriber.close0(cause);
+                return null;
+            });
+        }
+    }
+
+    @Override
+    public void abort() {
+        abort(AbortedStreamException.get());
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+
+        // `completionFuture` should be set before `SurroundingSubscriber` publishes data
+        // to guarantee the visibility of the abortion `cause` after
+        // SurroundingSubscriber is set in `subscriber0()`.
+        completionFuture.completeExceptionally(cause);
+
+        final SurroundingSubscriber<T> surroundingSubscriber = this.surroundingSubscriber;
+        if (surroundingSubscriber != null) {
+            surroundingSubscriber.close(cause);
+        }
+    }
+
+    static final class SurroundingSubscriber<T> implements Subscriber<T>, Subscription {
+
+        enum State {
+            REQUIRE_HEAD,
+            REQUIRE_BODY,
+            REQUIRE_TAIL,
+            REQUIRE_COMPLETE,
+            DONE,
+        }
+
+        private State state;
+
+        @Nullable
+        private final T head;
+        private final StreamMessage<T> publisher;
+        @Nullable
+        private final T tail;
+
+        private Subscriber<? super T> downstream;
+        private final EventExecutor executor;
+        @Nullable
+        private volatile Subscription upstream;
+
+        private long requested;
+        private long upstreamRequested;
```

Comment:
Q: Is it difficult to merge the two fields? It would be easier to track one count.

Reply:
> Is it difficult to merge the two fields? 

I believe so. It's not easy to track the two demands that are from downstream and to upstream with one field.

Reply:
> Not easy to track downstream and to upstream with one field.

Same with me. I really struggle to merge this two fields, but can't find any solution with this code structure 😅

Reply:
Okay. Let's keep it as is.